lib/order.rb

```diff
@@ -1,18 +1,90 @@
+require 'csv'
+require 'awesome_print'
+
+
 module Grocery
   class Order
     attr_reader :id, :products
 
+
     def initialize(id, products)
       @id = id
       @products = products
     end
 
-    def total
-      # TODO: implement total
+
+    def self.all
+      orders = []
+      # "This is a class method."
+      # This is a class method to access the class variabel
+      #returns the csv file to an array with headers
+      #orders_csv = CSV.read('support/orders.csv', 'r')
+      orders_csv = CSV.read("support/orders.csv", 'r',headers: true).to_a
+
+
+      orders_csv.each do |line|
+        orders << Grocery::Order.new(*line)
+      end
+      return orders
+    end
+
+
+    def self.find(id)
+      a = self.all
+
+      if id > a.length
+        return nil
+      end
+
+      if id >=1
+        specific_order = a[id-1]
+      elsif id < 0
+        specific_order = a[id]
+      else
+        return nil
+      end
+      return specific_order
+    end
+
+
+
+
+    def self.total
+      a = self.all
+      sum = 0
+      #put in a class level variable for total
+      if !(a.products.empty?) || a.products !=  nil
+        a.products.each do |product, cost|
+          sum += cost
+        end
+        sales_tax = 0.075
+        sum = (sum + (sum * sales_tax)).round(2)
+      end
+      return sum
+      # TODO: make changes where sum is returned
     end
 
+
     def add_product(product_name, product_price)
-      # TODO: implement add_product
+      a = self.all
+      if a.products.has_key? product_name
+        return false
```

Comment:
I'm not sure why you're calling `.all` inside this instance method. You're already inside of an `Order` here, so you shouldn't need the list of all orders for anything.

A correct version of this method would be:

```ruby
  def add_product(product_name, product_price)
      if @products.has_key? product_name
        return false
      else
        @products[product_name] = product_price
        return true
      end
    end
```

Looking at your commit history, you had this and then changed it. I'm not sure why.

Reply:
Switching this (and the other instance methods in this file) from `a.products` to `@products` made a bunch of your tests pass.

Reply:
I went back through the code and clarified for myself which were instance methods and which were class methods in comments. That seems to be where a lot of my confusion was coming from. I'm not sure why I need to instantiate the method before calling the instance method on it. I must have been testing it somewhere without instantiating the class itself and used that technique to make the output line up with what I thought it should be.